File: Baekjoon/Class3/33.2178 미로 탐색.cpp
#include <iostream>
#include <string>
#include <vector>
#include <queue>
using namespace std;

struct Point {
    int r, c;
};
constexpr Point MOVE_DIR[] = {
    {-1, 0},
    {1, 0},
    {0, -1},
    {0, 1}
};

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    int n, m;
    cin >> n >> m;

    vector<vector<int>> depths(n, vector<int>(m, 0)); // 벽 0
    for (int r = 0; r < n; r++) {
        string row;
        cin >> row;
        for (int c = 0; c < m; c++)
            if (row[c] == '1')
                depths[r][c] = -1; // 길 -1
    }

    queue<Point> bfs; // 너비 우선 탐색
    bfs.push({ 0, 0 }); // 출발 위치 { 0, 0 }
    depths[0][0] = 1; // 출발 깊이 1
    while (true) {
        const Point start = bfs.front();
        bfs.pop();
        if (start.r == n - 1 && start.c == m - 1) // 도착 위치 { n - 1, m - 1 }
            break;
        for (const Point& move_dir : MOVE_DIR) {
            int end_r = start.r + move_dir.r;
            int end_c = start.c + move_dir.c;
            if (end_r < 0 || end_r >= n || end_c < 0 || end_c >= m) // 경계 확인
                continue;
            if (depths[end_r][end_c] == -1) { // 이동
                bfs.push({ end_r, end_c });
                depths[end_r][end_c] = depths[start.r][start.c] + 1;
            }
        }
    }

    cout << depths[n - 1][m - 1]; // 도착 깊이
    return 0;
}

File: Baekjoon/Class3/33.2178 미로 탐색.java
import java.io.*;
import java.util.ArrayDeque;
import java.util.Deque;
import java.util.StringTokenizer;

class Main {
    private static class Point {
        int r, c;
        public Point(int r, int c) { this.r = r; this.c = c; }
    }
    private static final Point[] MOVE_DIR = {
            new Point(-1, 0),
            new Point(1, 0),
            new Point(0, -1),
            new Point(0, 1)
    };

    public static void main(String[] args) throws IOException {
        int n, m;
        int[][] depths;
        Deque<Point> bfs = new ArrayDeque<>();
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());

        depths = new int[n][m]; // 벽 0
        for(int r = 0; r < n; r++) {
            String row = br.readLine();
            for(int c = 0; c < m; c++)
                if(row.charAt(c) == '1')
                    depths[r][c] = -1; // 길 -1
        }

        bfs.addLast(new Point(0, 0)); // 출발 위치 (0, 0)
        depths[0][0] = 1; // 출발 깊이 1
        while(true) {
            Point start = bfs.removeFirst();
            if(start.r == n - 1 && start.c == m - 1) // 도착 위치 (n - 1, m - 1)
                break;
            for(Point moveDir: MOVE_DIR) {
                int endR = start.r + moveDir.r;
                int endC = start.c + moveDir.c;
                if(endR < 0 || endR >= n || endC < 0 || endC >= m) // 경계 확인
                    continue;
                if(depths[endR][endC] == -1) { // 이동
                    bfs.addLast(new Point(endR, endC));
                    depths[endR][endC] = depths[start.r][start.c] + 1;
                }
            }
        }

        System.out.print(depths[n - 1][m - 1]); // 도착 깊이
        br.close();
    }
}